source/Web/Features/Statistics/StatisticsRepository.cs

```diff
@@ -1,125 +1,79 @@
-    {
-        Task AddStatisticsForResult(InvestigationResult result);
-        IReadOnlyList<PackageStatistic> GetAllPackageStatistics();
-        void UpdateSupportTypeFor(PackageStatistic stat, SupportType supportType);
-    }
-
     public class StatisticsRepository : IStatisticsRepository
     {
         // Only log packages found on Nuget.org
         private static readonly SupportType[] AddStatisticsFor = { SupportType.Unsupported, SupportType.Supported, SupportType.PreRelease };
 
-        private readonly string _connectionString;
+        private readonly Func<AppDbContext> _contextFactory;
 
-        public StatisticsRepository(IConfigurationRoot configuration)
+        public StatisticsRepository(Func<AppDbContext> contextFactory)
         {
-            _connectionString = configuration["ConnectionString"];
-            if (string.IsNullOrWhiteSpace(_connectionString))
-                throw new Exception("The configuration setting 'ConnectionString' is not set");
+            _contextFactory = contextFactory;
         }
 
-        public async Task AddStatisticsForResult(InvestigationResult result)
+        public async Task AddStatisticsForResultAsync(InvestigationResult result, CancellationToken cancellationToken)
         {
             try
             {
-                using (var con = new SqlConnection(_connectionString))
-                {
-                    await con.OpenAsync();
-                    var tasks = result.GetAllDistinctRecursive()
-                        .Where(p => AddStatisticsFor.Contains(p.SupportType))
-                        .Select(p => AddStatistic(con, p))
-                        .ToArray();
+                var tasks = result.GetAllDistinctRecursive()
+                    .Where(p => AddStatisticsFor.Contains(p.SupportType))
+                    .Select(p => AddStatisticAsync(p, cancellationToken));
 
-                    await Task.WhenAll(tasks);
-                }
+                await Task.WhenAll(tasks);
             }
             catch (Exception ex)
             {
                 Log.Logger.Error(ex, "Exception writing statistics");
             }
         }
 
-        private async Task AddStatistic(SqlConnection con, PackageResult package)
+        private async Task AddStatisticAsync(PackageResult package, CancellationToken cancellationToken)
         {
-            const string sql = @"MERGE dbo.PackageStatistics AS t
-USING
-(
-    SELECT @name as 'Name', @latestSupportType as 'LatestSupportType'
-) AS s
-ON t.Name = s.Name
-
-WHEN MATCHED THEN
-    UPDATE SET LatestSupportType = s.LatestSupportType, [Count] = [Count] + 1 
-
-WHEN NOT MATCHED THEN
-    INSERT (Name, LatestSupportType, Count)
-    VALUES (s.Name, s.LatestSupportType, 1);";
-
-            try
+            await using var context = _contextFactory();
+            var packageStatistic = await context.PackageStatistics.FindAsync(new object[] {package.PackageName}, cancellationToken);
```

Comment:
This is no longer atomic. I used the `MERGE` command as it guaranteed (from my understanding) atomic insert/update of that record. With this change two threads could update the same statistic record and only the last in wins.

Reply:
Good catch! I addressed this issue in 49561dba1e6f68497d69efe13669c1c8e0e0b6d1. I used [FlexLabs.EntityFrameworkCore.Upsert](https://github.com/artiomchi/FlexLabs.Upsert) as EF Core does not handle upsert out of the box.

By increasing the EF Core log level to information (`.MinimumLevel.Override("Microsoft.EntityFrameworkCore", LogEventLevel.Information)`) and enabling sensitive data logging (`optionsBuilder.EnableSensitiveDataLogging();`) we can see how FlexLabs.Upsert works:

With SQLite:

```sql
-- Executing DbCommand [Parameters=[@p0='Newtonsoft.Json' (Nullable = false) (Size = 15), @p1='1' (DbType = String), @p2='Supported' (Nullable = false) (Size = 9), @p3='1' (DbType = String)], CommandType='Text', CommandTimeout='30']
INSERT INTO "PackageStatistics" AS "T" ("Name", "Count", "LatestSupportType") VALUES (@p0, @p1, @p2) ON CONFLICT ("Name") DO UPDATE SET "Count" = ( "T"."Count" + @p3 )
```

With SQL Server:

```sql
-- Executing DbCommand [Parameters=[@p0='Newtonsoft.Json' (Nullable = false) (Size = 450), @p1='1', @p2='Supported' (Nullable = false) (Size = 4000), @p3='1'], CommandType='Text', CommandTimeout='30']
MERGE INTO [PackageStatistics] WITH (HOLDLOCK) AS [T] USING ( VALUES (@p0, @p1, @p2) ) AS [S] ([Name], [Count], [LatestSupportType]) ON [T].[Name] = [S].[Name] WHEN NOT MATCHED BY TARGET THEN INSERT ([Name], [Count], [LatestSupportType]) VALUES ([Name], [Count], [LatestSupportType]) WHEN MATCHED THEN UPDATE SET [Count] = ( [T].[Count] + @p3 );
```